File: src/main/python/readwrite/pace_extended.py
import networkx as nx
from typing import TextIO, Optional

__all__ = [
    'read_pace_extended',
    'load_pace_extended',
    'write_pace_extended',
    'save_pace_extended',
]


def read_pace_extended(input: TextIO) -> tuple[nx.Graph, Optional[int], Optional[int]]:
    G = nx.Graph()
    for line in input.readlines():
        line = line.strip()
        if not line:
            continue
        if line.startswith('c'):
            continue  # ignore comments

        tokens = line.split()

        if tokens[0] == 'p':
            assert tokens[1] == 'twwe', f'unexpected problem descriptor: {tokens[1]}'
            assert 4 <= len(tokens) <= 6

            given_n = int(tokens[2])
            given_m = int(tokens[3])
            given_ub = int(tokens[4]) if len(tokens) >= 5 else None
            given_lb = int(tokens[5]) if len(tokens) >= 6 else None

        elif tokens[0] == 'v':
            assert len(tokens) == 2

            v = int(tokens[1])
            G.add_node(v)

        elif tokens[0] == 'e':
            assert 3 <= len(tokens) <= 4

            u = int(tokens[1])
            v = int(tokens[2])
            c = int(tokens[3]) if len(tokens) >= 4 else 0  # defaults to black edge (0)
            G.add_edge(u, v, color=c)

        else:
            assert False, f'unexpected line: {line}'

    assert given_n == G.number_of_nodes(), 'inconsistent n'
    assert given_m == G.number_of_edges(), 'inconsistent m'
    return G, given_ub, given_lb


def load_pace_extended(path: str) -> tuple[nx.Graph, Optional[int], Optional[int]]:
    with open(path) as f:
        return read_pace_extended(f)


def write_pace_extended(output: TextIO, G: nx.Graph, comments: list[str] = [], ub: Optional[int] = None, lb: Optional[int] = None, offset: int=0) -> None:
    problem = 'twwe'  # problem descriptor (TWin-Width Extended)
    n = G.number_of_nodes()
    m = G.number_of_edges()

    if ub is None and lb is not None:
        ub = n

    # comments
    for comment in comments:
        output.write(f'c {comment}\n')

    # p-line
    tokens = ['p', problem, str(n), str(m)] + ([] if ub is None else [str(ub)]) + ([] if lb is None else [str(lb)])
    output.write(' '.join(tokens) + '\n')

    # vertices (not necessarily consecutive)
    for v in G.nodes():
        output.write(f'v {v + offset}\n')

    # colored edges
    for u, v in G.edges():
        color = G.edges[u, v]['color']  # 0 or 1
        output.write(f'e {u + offset} {v + offset} {color}\n')


def save_pace_extended(path: str, G: nx.Graph, comments: list[str] = []) -> None:
    with open(path, 'w') as f:
        write_pace_extended(f, G, comments)

File: src/main/cpp/readwrite/pace_extended.cpp
#include <fstream>

#include "pace_extended.hpp"
#include "util/logger.hpp"

namespace readwrite {
ds::ProblemInstance read_pace_extended(std::istream &is) {
  typedef ds::graph::TriGraph::VertexList VertexList;
  typedef ds::graph::TriGraph::ColoredEdgeList ColoredEdgeList;

  VertexList vertices;
  ColoredEdgeList edges;

  int n = -1;
  int m = -1;
  int lb = 0;
  int ub = -1;
  int i = 0;

  for (std::string line; std::getline(is, line);) {
    if (line.empty()) continue;

    while (i < static_cast<int>(line.size()) && !std::isprint(line[i])) ++i;
    if (line[i] == 'c') continue;  // ignore comments

    auto tokens = util::split(line);
    if (tokens[0] == "p") {  // p-line
      if (tokens.size() < 4) throw std::invalid_argument(util::format("unexpected p-line: %s", line.c_str()));
      if (tokens[1] != "twwe") throw std::invalid_argument(util::format("unexpected p-line: %s", line.c_str()));

      n = std::stoi(tokens[2]);
      m = std::stoi(tokens[3]);
      if (tokens.size() >= 5) ub = std::stoi(tokens[4]);
      if (tokens.size() >= 6) lb = std::stoi(tokens[5]);

      if (lb >= n)
        if (ub >= 0 && ub < lb) {
          throw std::invalid_argument(util::format("upper bound too small: n=%d, lb=%d, ub=%d", n, lb, ub));
        }
    } else if (tokens[0] == "v") {  // v-line
      if (tokens.size() != 2) throw std::invalid_argument("unexpected v-line");

      auto v = std::stoi(tokens[1]);
      vertices.push_back(v);
    } else if (tokens[0] == "e") {  // e-line
      if (tokens.size() != 4) throw std::invalid_argument("unexpected e-line");

      auto u = std::stoi(tokens[1]);
      auto v = std::stoi(tokens[2]);
      auto color = std::stoi(tokens[3]);
      if (color != ds::graph::TriGraph::Black && color != ds::graph::TriGraph::Red) {
        throw std::invalid_argument(util::format("unexpected color: %d", color));
      }
      edges.push_back({{u, v}, color});
    }
  }

  ds::graph::TriGraph G(vertices, edges);

  // verification
  if (n != static_cast<int>(G.number_of_vertices())) throw std::invalid_argument("inconsistent n");
  if (m != static_cast<int>(G.number_of_edges())) throw std::invalid_argument("inconsistent m");
  if (ub >= 0 && lb > ub) {
    throw std::invalid_argument(util::format("lower bound larger than upper bound: n=%d, lb=%d, ub=%d", n, lb, ub));
  }

  ds::ProblemInstance ret = {G, lb, ub};
  return ret;
}

ds::ProblemInstance load_pace_extended(char const *path) {
  std::ifstream f(path);
  if (f.fail()) throw std::invalid_argument(util::format("Failed to open file: %s", path));
  return read_pace_extended(f);
}
}  // namespace readwrite
